### PR TITLE
feat: ship binaries with the extension

### LINKED ISSUE #71: Feature Request: Optionally install the new official precompiled binaries for macOS and Linux aarch64 from shellcheck >0.7.0
https://github.com/koalaman/shellcheck/blob/master/CHANGELOG.md
```
v0.7.0 - 2019-07-28
Added

    Precompiled binaries for macOS and Linux aarch64
```
Could be pretty cool :) Would be good to make it a clear option though as some Linux users will only want to use their distro's binaries. Maybe it detects if you have shellcheck in the path and says something like:

"shellcheck version xxx detected on your machine, would you like to use the bundled version xxx instead?"
"you don't have any version of shellcheck installed, would you like to use the bundled version xxx?"
"you are using windows. this is 2019, are you insane? please install linux" :)

Thanks for this fantastic extension!

### PR DESCRIPTION
**Pull request by felipecrs**
The binaries are downloaded with [`bindl`](https://github.com/felipecrs/bindl), which I just wrote. It basically only requires a config file, so take a look at the `bindl.config.js`.

- [X] Download the binaries when packaging the extension
- [X] Set the default configuration for `executablePath` as empty and use the builtin binaries if it's empty 

[shellcheck-0.9.0.vsix.zip](https://github.com/timonwong/vscode-shellcheck/files/5113578/shellcheck-0.9.0.vsix.zip)

Fixes #71

### CODE REVIEW

**Comment by felipecassiors** [comment]
At this time, the extension is picking the binary shipped with the extension in case we are running in a supported platform (Linux, MacOS or Windows).

nUDAH8atxV

By the way, shipping the binaries with the extension would add an extra 8MB in its size (not so big). 

I'm wondering how can I make the desired priority to be respected. Any suggestion, @timonwong?

**Comment by timonwong** [comment]
@felipecassiors I'm just worried about the legal issues about GPLv3, since this is a distribution of GPLv3 binary, should we distribute the source of shellcheck as well (Or provide user enough information about how to fetch & build the binary) ?

**Comment by felipecassiors** [comment]
> @felipecassiors I'm just worried about the legal issues about GPLv3, since this is a distribution of GPLv3 binary, should we distribute the source of shellcheck as well (Or provide user enough information about how to fetch & build the binary) ?

The second option is certainly better, I'll take a look at the feasibility.

**Comment by timonwong** [comment]
@felipecassiors I think include a README and a GPL license in the binaries/ directory should be enough

And can you make this optional (and by default false)? or it may break on those machines with customized wrapper script or WSL...

**Comment by felipecassiors** [comment]
> And can you make this optional (and by default false)? or it may break on those machines with customized wrapper script or WSL...

I believe the way forward is to make it default to make it easier for newcomers to get started without needing any configuration. Why do you think it will break? Even for Windows users using WSL, we will ship binaries for Windows as well, and Linux, so both people working from inside or outside WSL would get `shellcheck` working.

I know this might be a major change for existing users, but I don't how this would affect negatively them. For people that really want to use "customized" binaries or specific versions, they still have the option to configure which binary to use.

**Comment by felipecassiors** [comment]
For this PR, I'm working on an NPM module which will download the binaries automatically during the `vsce` packaging.

**Comment by felipecassiors** [comment]
I see [`shell-format`](https://marketplace.visualstudio.com/items?itemName=foxundermoon.shell-format) shipping binaries included, and optionally you can specify a custom path for the binary in case you don't want to use the builtin. 

[`bash-debug`](https://marketplace.visualstudio.com/items?itemName=rogalmic.bash-debug) on the other hand, also ship binaries with the extension, but does not give the user the ability to set a custom binary. Most likely because it wants to ensure a certain version that is known to work with the extension.

I'll try to figure out more similar extensions.

**Comment by beltaurus** [comment]
### TL;DR
- GPLv3 is no obstacle.
- If we're going for convenience, then ship the 3rd party executables, but keep the configuration/settings simple.
- Convenience is not always possible (e.g. remote applications).

### Thoughts
I haven't read the details of the back and forth, but wanted to address some of the issues that seem to pop up:
With GPLv3, you don't _have_ to include the source code of binaries you distribute, if you _make the respective source code available in a clear and obvious way_. That means that during the installation of your extension, you need to point to a location on YOUR distribution medium, where the source code corresponding to the distribution you include is available.
Disclaimer: I am not a lawyer, but these are the instructions I have received before with regards to this topic.

Next: I don't know if it is always necessary to include those binaries in the first place... 
An example is actually for remote applications: on Linux, I already have the shellcheck installed. And for other languages, I have the corresponding checkers installed for stand-alone usage (e.g. PerlCritic for Perl). And therefore, extensions that use these applications, don't have to include it in the distribution.
So what I typically look at: how easy is it to install those third parties, without risking to become an obstacle for getting your extension to work? 
That aside, convenience is king, then including the required 3rd party binaries would indeed be nice. And as mentioned above, it provides some extra robustness, since we test what we distribute, right?! Yet, if convenience is king, than I would also not complicate the situation by making the 3rd party application "selectable". Keep the settings simple and to just a few...
E.g. If the 3rd party application has configuration files (`~/.appicationrc`), there is not really a need to have your extension GUI control the configuration...
But really think what that means! On Windows, I more often than not run in WSL, where the applications are to be installed visible to the Bash shell... Make sure that doesn't clash with installations that might already be there...

**Comment by timonwong** [comment]
@beltaurus Agreed, so I think we shoud:
1. Ship with the binaries, GPL license and readme describe how to fetch shellcheck source (github project link is enough)
2. Make `shellcheck.executable` optional, when unset, use our bundled version
2.1. Better, we can prompt our users (using a dialog or open a markdown documentation), after upgrading, encourage them to use the bundle version?
3. Make sure our bundled shellcheck works on most machines (OS/arch)
3.1. When error (ABI incompatibility for example), prompts our users to download (or build) shellcheck manually.

We can release some testing versions gathering the feedback, It's not necessary to do all the things in just one PR.
Any thoughts?

**Comment by felipecassiors** [comment]
> Any thoughts?

Sounds great! I'm working (and I didn't have time to finish it this weekend) on an NPM project, which can automatically download binaries from compressed files and extract only the files we might want to a certain folder, according to a configuration file. 

However, there is no need to have such an automated process in the beginning.

**Review by timonwong — APPROVED** [review]
LGTM
